src/box/memtx_tx.c

```diff
@@ -365,23 +590,123 @@ memtx_tx_handle_conflict(struct txn *breaker, struct txn *victim)
+ * was replaced from primary key and that is why @a story
+ * is the only reason why the tuple cannot be deleted.
+ */
+static inline void
+memtx_tx_story_pin(struct memtx_story *story)
+{
+	assert(!story->is_pinned);
+	assert(story->pin_status < MEMTX_TX_STORY_MAX);
+
+	story->is_pinned = true;
+	struct memtx_tx_stats *stats = &txm.tuple_pin_stats[story->pin_status];
+	size_t tuplesize = tuple_size(story->tuple);
+	memtx_tx_stats_collect(stats, tuplesize);
+}
+
+/**
+ * Notify memory manager that a tuple referenced by @a story
+ * was placed to primary key.
+ */
+static inline void
+memtx_tx_story_unpin(struct memtx_story *story)
+{
+	assert(story->is_pinned);
+	assert(story->pin_status < MEMTX_TX_STORY_MAX);
+
+	story->is_pinned = false;
+	struct memtx_tx_stats *stats = &txm.tuple_pin_stats[story->pin_status];
+	size_t tuplesize = tuple_size(story->tuple);
+	memtx_tx_stats_discard(stats, tuplesize);
+}
+
+static inline bool
+memtx_tx_story_is_pinned(struct memtx_story *story)
+{
+	return story->is_pinned;
+}
+
+/** Set status of story (see MEMTX_TX_PIN_TYPE) */
+static inline void
+memtx_tx_story_set_status(struct memtx_story *story,
+			  enum memtx_tx_story_status new_status)
+{
+	assert(story->pin_status < MEMTX_TX_STORY_MAX);
+	enum memtx_tx_story_status old_status = story->pin_status;
+	if (old_status == new_status)
+		return;
+	story->pin_status = new_status;
+	struct memtx_tx_stats *old_story_stats = &txm.story_stats[old_status];
+	struct memtx_tx_stats *new_story_stats = &txm.story_stats[new_status];
+	size_t story_size = memtx_story_size(story);
+	memtx_tx_stats_discard(old_story_stats, story_size);
+	memtx_tx_stats_collect(new_story_stats, story_size);
+	if (story->is_pinned) {
+		size_t tuplesize = tuple_size(story->tuple);
+		struct memtx_tx_stats *old = &txm.tuple_pin_stats[old_status];
+		struct memtx_tx_stats *new = &txm.tuple_pin_stats[new_status];
+		memtx_tx_stats_discard(old, tuplesize);
+		memtx_tx_stats_collect(new, tuplesize);
+	}
+}
+
+/**
+ * Use this method to ref tuple that belongs to @a story
+ * by primary index. Do not use bare tuple_ref!!!
+ */
+static inline void
+memtx_tx_ref_to_primary(struct memtx_story *story)
+{
+	assert(story != NULL);
+	assert(memtx_tx_story_is_pinned(story));
+	tuple_ref(story->tuple);
+	memtx_tx_story_unpin(story);
+}
+
+/**
+ * Use this method to unref tuple that belongs to @a story
+ * from primary index. Do not use bare tuple_unref!!!
+ */
+static inline void
+memtx_tx_unref_from_primary(struct memtx_story *story)
+{
+	assert(story != NULL);
+	assert(!memtx_tx_story_is_pinned(story));
+	tuple_unref(story->tuple);
+	memtx_tx_story_pin(story);
+}
+
 /**
  * Create a new story and link it with the @a tuple.
  * @return story on success, NULL on error (diag is set).
  */
 static struct memtx_story *
-memtx_tx_story_new(struct space *space, struct tuple *tuple)
+memtx_tx_story_new(struct space *space, struct tuple *tuple, bool pin)
```

Comment:
Without any comment it's hard to understand that `pin` arg means.

Reply:
Renamed the arg